### DISCUSSION TITLE
controlnet error

### DISCUSSION THREAD

**Discussion by hoodady** [Panchovix/stable-diffusion-webui-reForge#38] [General]
[text2prompt] Following databases are available:
    all-mpnet-base-v2 : danbooru_strict
Loading model jibMixRealisticXL_v120TheLastHuzzah.safetensors [23a0978641] (1 out of 1)
Loading weights [23a0978641] from Z:\Stable Forge\stable-diffusion-webui-forge\models\Stable-diffusion\jibMixRealisticXL_v120TheLastHuzzah.safetensors
*** module 'lib_controlnet.external_code' has no attribute 'get_models' ***
    Traceback (most recent call last):
      File "Z:\Stable Forge\stable-diffusion-webui-forge\extensions\sd-webui-replacer\replacer\extensions\controlnet.py", line 213, in getInpaintModels
        models = external_code.get_models()
    AttributeError: module 'lib_controlnet.external_code' has no attribute 'get_models'

---
*** module 'lib_controlnet.external_code' has no attribute 'get_models' ***
    Traceback (most recent call last):
      File "Z:\Stable Forge\stable-diffusion-webui-forge\extensions\sd-webui-replacer\replacer\extensions\controlnet.py", line 213, in getInpaintModels
        models = external_code.get_models()
    AttributeError: module 'lib_controlnet.external_code' has no attribute 'get_models'

**Comment by Panchovix** [comment]
Can you share the complete error in the console? Maybe something is not updated on external_code.py (if on dev_upstream)

**Reply by hoodady** [reply]
venv "Z:\Stable Forge\stable-diffusion-webui-forge\venv\Scripts\Python.exe"
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Version: f0.0.20.1-v1.10.0RC-latest-543-g39e8386d
Commit hash: 39e8386d639d9624259610c461467f62e87a631e
CUDA 12.1
Launching Web UI with arguments: --api --cuda-malloc --xformers
Using cudaMallocAsync backend.
Total VRAM 6144 MB, total RAM 15773 MB
xformers version: 0.0.23.post1
Set vram state to: NORMAL_VRAM
Device: cuda:0 NVIDIA GeForce RTX 3060 Laptop GPU : cudaMallocAsync
Hint: your device supports --pin-shared-memory for potential speed improvements.
Hint: your device supports --cuda-stream for potential speed improvements.
VAE dtype: torch.bfloat16
CUDA Stream Activated:  False
2024-07-21 12:07:28.395651: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-07-21 12:07:29.153428: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
Using xformers cross attention
ControlNet preprocessor location: Z:\Stable Forge\stable-diffusion-webui-forge\models\ControlNetPreprocessor
CHv1.8.10: Get Custom Model Folder
[-] ADetailer initialized. version: 24.6.0, num models: 37
12:07:48 - ReActor - STATUS - Running v0.7.0-a2 on Device: CUDA
[text2prompt] Following databases are available:
    all-mpnet-base-v2 : danbooru_strict
Loading model rawcharmAmateur_v10VAE.safetensors [321ed9246e] (1 out of 1)
Loading weights [321ed9246e] from Z:\Stable Forge\stable-diffusion-webui-forge\models\Stable-diffusion\rawcharmAmateur_v10VAE.safetensors
*** module 'lib_controlnet.external_code' has no attribute 'get_models' ***
    Traceback (most recent call last):
      File "Z:\Stable Forge\stable-diffusion-webui-forge\extensions\sd-webui-replacer\replacer\extensions\controlnet.py", line 213, in getInpaintModels
        models = external_code.get_models()
    AttributeError: module 'lib_controlnet.external_code' has no attribute 'get_models'

---
*** module 'lib_controlnet.external_code' has no attribute 'get_models' ***
    Traceback (most recent call last):
      File "Z:\Stable Forge\stable-diffusion-webui-forge\extensions\sd-webui-replacer\replacer\extensions\controlnet.py", line 213, in getInpaintModels
        models = external_code.get_models()
    AttributeError: module 'lib_controlnet.external_code' has no attribute 'get_models'

---
CHv1.8.10: Set Proxy:
model_type EPS
UNet ADM Dimension 2816
2024-07-21 12:07:57,732 - ControlNet - INFO - ControlNet UI callback registered.
[text2prompt] Loading database with name "all-mpnet-base-v2 : danbooru_strict"...
[text2prompt] Database loaded
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 76.9s (prepare environment: 30.6s, import torch: 5.7s, import gradio: 1.0s, setup paths: 3.3s, other imports: 0.8s, list SD models: 0.5s, load scripts: 20.8s, scripts before_ui_callback: 0.2s, create ui: 6.1s, gradio launch: 0.9s, add APIs: 5.8s, app_started_callback: 1.0s).
Using xformers attention in VAE
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
Using xformers attention in VAE
extra {'cond_stage_model.clip_l.text_projection', 'cond_stage_model.clip_l.logit_scale', 'cond_stage_model.clip_g.transformer.text_model.embeddings.position_ids'}
loaded straight to GPU
To load target model SDXL
Begin to load 1 model
[Memory Management] Current Free GPU Memory (MB) =  196.95165252685547
[Memory Management] Model Memory (MB) =  0.03814697265625
[Memory Management] Minimal Inference Memory (MB) =  1024.0
[Memory Management] Estimated Remaining GPU Memory (MB) =  -827.0864944458008
[Memory Management] Requested SYNC Preserved Memory (MB) =  0.0
[Memory Management] Parameters Loaded to SYNC Stream (MB) =  4897.0483474731445
[Memory Management] Parameters Loaded to GPU (MB) =  0.0
Moving model(s) has taken 5.04 seconds
To load target model SDXLClipModel
Begin to load 1 model
[Memory Management] Current Free GPU Memory (MB) =  5073.620208740234
[Memory Management] Model Memory (MB) =  2144.3546981811523
[Memory Management] Minimal Inference Memory (MB) =  1024.0
[Memory Management] Estimated Remaining GPU Memory (MB) =  1905.265510559082
Moving model(s) has taken 2.92 seconds
Model loaded in 29.1s (load weights from disk: 0.8s, forge load real models: 22.9s, load textual inversion embeddings: 1.0s, calculate empty prompt: 4.1s).

**Reply by hoodady** [reply]
im on main i think

**Reply by Panchovix** [reply]
The error comes before loading the UI? I think that comes from the extension "sd-webui-replacer". It is searching get_models on lib_controlnet.external_code (on the default controlnet extension) while on forge/reforge it is located on ldm_patched.modules.controlnet

**Reply by hoodady** [reply]
ok I removed the replacer, and the error is gone cheers mate
